go/service/chat_boxer.go

```diff
@@ -186,43 +275,56 @@ func (b *chatBoxer) signJSON(data interface{}, kp libkb.NaclSigningKeyPair, pref
 	return b.sign(encoded, kp, prefix)
 }
 
-func exportSigInfo(si *libkb.NaclSigInfo) chat1.SignatureInfo {
-	return chat1.SignatureInfo{
-		V: si.Version,
-		S: si.Sig[:],
-		K: si.Kid,
-	}
-}
-
 // sign signs msg with a NaclSigningKeyPair, returning a chat1.SignatureInfo.
 func (b *chatBoxer) sign(msg []byte, kp libkb.NaclSigningKeyPair, prefix libkb.SignaturePrefix) (chat1.SignatureInfo, error) {
 	sig, err := kp.SignV2(msg, prefix)
 	if err != nil {
 		return chat1.SignatureInfo{}, err
 	}
-	return exportSigInfo(sig), nil
+	sigInfo := chat1.SignatureInfo{
+		V: sig.Version,
+		S: sig.Sig[:],
+		K: sig.Kid,
+	}
+	return sigInfo, nil
 }
 
-// verifyMessageBoxed verifies the header and body signatures in a boxed
-// message.
-func (b *chatBoxer) verifyMessageBoxed(msg chat1.MessageBoxed) error {
-	header, err := json.Marshal(msg.ClientHeader)
+// verifyMessage checks that a message is valid.
+func (b *chatBoxer) verifyMessage(header keybase1.HeaderPlaintext, msg chat1.MessageBoxed) error {
+	headerVersion, err := header.Version()
 	if err != nil {
 		return err
 	}
-	if !b.verify(header, msg.HeaderSignature, libkb.SignaturePrefixChatHeader) {
-		return libkb.BadSigError{E: "header signature invalid"}
+
+	switch headerVersion {
+	case keybase1.HeaderPlaintextVersion_V1:
+		return b.verifyMessageHeaderV1(header.V1(), msg)
+	default:
+		return libkb.NewChatHeaderVersionError(headerVersion)
 	}
+}
 
-	if !b.verify(msg.BodyCiphertext.E, msg.BodySignature, libkb.SignaturePrefixChatBody) {
-		return libkb.BadSigError{E: "body signature invalid"}
+// verifyMessageHeaderV1 checks the body hash, header signature, and signing key validity.
+func (b *chatBoxer) verifyMessageHeaderV1(header keybase1.HeaderPlaintextV1, msg chat1.MessageBoxed) error {
```

Comment:
We should have tests for all of the error cases in this function, since it's so security sensitive.